lib/widgets/compose_box.dart

```diff
@@ -234,61 +307,110 @@ class _AttachFileButton extends StatelessWidget {
+  }
+}
+
+class _AttachMediaButton extends StatelessWidget {
+  const _AttachMediaButton({required this.contentController, required this.contentFocusNode});
+
+  final ContentTextEditingController contentController;
+  final FocusNode contentFocusNode;
+
+  _handlePress(BuildContext context) async {
+    FilePickerResult? result;
+    try {
+      result = await FilePicker.platform.pickFiles(type: FileType.media, allowMultiple: true, withReadStream: true);
+    } catch (e) {
+      // TODO(i18n)
+      showErrorDialog(context: context, title: 'Error', message: e.toString());
+      return;
+    }
+    if (result == null) {
+      return; // User cancelled; do nothing
     }
 
-    if (tooLargeFiles.isNotEmpty) {
-      final listMessage = tooLargeFiles
-        .map((file) => '${file.name}: ${(file.size / (1 << 20)).toStringAsFixed(1)} MiB')
-        .join('\n');
-      showErrorDialog( // TODO(i18n)
-        context: context,
-        title: 'File(s) too large',
-        message:
-          '${tooLargeFiles.length} file(s) are larger than the server\'s limit of ${store.maxFileUploadSizeMib} MiB and will not be uploaded:\n\n$listMessage');
+    if (context.mounted) {} // https://github.com/dart-lang/linter/issues/4007
+    else {
+      return;
     }
 
-    final List<(int, PlatformFile)> uploadsInProgress = [];
-    for (final file in rightSizeFiles) {
-      final tag = contentController.registerUploadStart(file.name);
-      uploadsInProgress.add((tag, file));
+    final Iterable<_File> files = result.files.map((f) {
+      assert(f.readStream != null);  // We passed `withReadStream: true` to pickFiles.
+      return _File(content: f.readStream!, length: f.size, filename: f.name);
+    });
+    _uploadFiles(context: context, contentController: contentController, contentFocusNode: contentFocusNode,
+        files: files);
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return IconButton(icon: const Icon(Icons.image), onPressed: () => _handlePress(context));
+  }
+}
+
+class _AttachFromCameraButton extends StatelessWidget {
+  _AttachFromCameraButton({required this.contentController, required this.contentFocusNode});
+
+  final _picker = ImagePicker();
+  final ContentTextEditingController contentController;
+  final FocusNode contentFocusNode;
+
+  _handlePress(BuildContext context) async {
+    final XFile? result;
+    try {
+      // Ideally we'd open a platform interface that lets you choose between
+      // taking a photo and a video. `image_picker` doesn't yet have that
+      // option: https://github.com/flutter/flutter/issues/89159
+      // so just stick with images for now. We could add another button for
+      // videos, but we don't want too many buttons.
+      result = await _picker.pickImage(source: ImageSource.camera, requestFullMetadata: false);
+    } catch (e) {
+      if (e is PlatformException && e.code == 'camera_access_denied') {
```

Comment:
Could one imagine a nice language feature that lets you match on more than just the type of the thing being thrown? The following is already possible:

```dart
} on PlatformException catch (e) {
```

But I'm thinking like, I dunno,

```
} on PlatformException(code: var c) when c == 'camera_access_denied' catch (e) {
```

or something?

This may be relevant: https://github.com/dart-lang/language/issues/112#issuecomment-1021530485